kernel/task/src/lib.rs

```diff
@@ -678,13 +678,59 @@ impl Task {
     }
 
     /// Blocks this `Task` by setting its runstate to [`RunState::Blocked`].
-    pub fn block(&self) {
-        self.runstate.store(RunState::Blocked);
+    ///
+    /// Returns the previous runstate. Trying to block an exited or reaped task
+    /// will return an error.
+    pub fn block(&self) -> Result<RunState, ()> {
+        if self
+            .runstate
+            .compare_exchange(RunState::Initing, RunState::Blocked)
+            .is_ok()
+        {
+            Ok(RunState::Initing)
+        } else if self
+            .runstate
+            .compare_exchange(RunState::Runnable, RunState::Blocked)
+            .is_ok()
+        {
+            Ok(RunState::Runnable)
+        } else if self
+            .runstate
+            .compare_exchange(RunState::Blocked, RunState::Blocked)
+            .is_ok()
+        {
+            Ok(RunState::Blocked)
+        } else {
+            Err(())
+        }
```

Comment:
IMO these should be re-ordered to move the common success case to the top, where the current runstate is `Runnable` and the new one is `Blocked`. That way we're not always wasting cycles on an expensive atomic conditional for a case (`Initing` --> `Blocked`) that will likely never occur.

I think it should be an error to attempt to block a task that is in the `Initing` runstate, since that means something went wrong with initialization and you were able to obtain a reference to a task that wasn't yet finished being initialized. I don't believe that could happen, but better safe than sorry.
For the one case where this is currently accepted (spawning a task that shouldn't run), perhaps we can add a separate code path to allow that transition in only this case.

Could you also condense the formatting a bit?

Reply:
Also should it be considered an error to attempt to block a task that is already blocked? Might be indicative of a logic error in the calling application; currently it's fine and doesn't hurt, but still worth raising the question.

Reply:
By separate code path do you mean a function?

> Also should it be considered an error to attempt to block a task that is already blocked?

Maybe? The only legitimate use case I can currently think of is the shell processing a Ctrl+Z signal while running a task that is already blocked by something (e.g. Mutex). But in my rework of the `tty`, I add a separate `Suspended` task state similar to Linux's `STOPPED` which ensures the two states aren't mixed.

Reply:
> By separate code path do you mean a function?

yes, sorry, i meant exposing a new dedicated function within `task` for this purpose alone, i.e., something like `block_new_task()` or `block_initing_task()` which only allows a Runstate transition from `Initing` --> `Blocked`.

Reply:
As for blocking an already-blocked task, for now let's just log a warning and see where/when it occurs.